File: src/main/java/org/stt/gui/jfx/text/CommandHighlighter.java
package org.stt.gui.jfx.text;

import org.antlr.v4.runtime.CharStream;
import org.antlr.v4.runtime.CommonTokenStream;
import org.antlr.v4.runtime.Token;
import org.antlr.v4.runtime.TokenStream;
import org.antlr.v4.runtime.tree.ParseTree;
import org.antlr.v4.runtime.tree.TerminalNode;
import org.fxmisc.richtext.StyleClassedTextArea;
import org.stt.command.CaseInsensitiveInputStream;
import org.stt.config.ReportConfig;
import org.stt.grammar.EnglishCommandsBaseVisitor;
import org.stt.grammar.EnglishCommandsLexer;
import org.stt.grammar.EnglishCommandsParser;
import org.stt.grammar.EnglishCommandsVisitor;
import org.stt.text.ItemGrouper;

import javax.inject.Inject;
import java.util.Arrays;
import java.util.Collections;
import java.util.Objects;

public class CommandHighlighter {
    private final ItemGrouper itemGrouper;
    private final ReportConfig config;
    private StyleClassedTextArea textArea;
    private EnglishCommandsVisitor<Void> visitor = new Highlighter();

    public CommandHighlighter(ItemGrouper itemGrouper,
                              ReportConfig config,
                              StyleClassedTextArea styleClassedTextArea) {
        this.itemGrouper = itemGrouper;
        this.config = config;
        this.textArea = Objects.requireNonNull(styleClassedTextArea);
    }

    public void update() {
        if (textArea.getLength() == 0) {
            return;
        }
        textArea.clearStyle(0, textArea.getLength());
        String text = textArea.getText();
        CharStream input = new CaseInsensitiveInputStream(text);
        EnglishCommandsLexer lexer = new EnglishCommandsLexer(input);
        TokenStream tokenStream = new CommonTokenStream(lexer);
        EnglishCommandsParser parser = new EnglishCommandsParser(tokenStream);
        parser.command().accept(visitor);
        ItemGrouper.Group[] groups = itemGrouper.getGroupsOf(text).toArray(new ItemGrouper.Group[0]);
        for (int i = 0; i < groups.length - 1; i++) {
            ItemGrouper.Group group = groups[i];
            if (group.type == ItemGrouper.Type.MATCH) {
                textArea.setStyle(group.range.start, group.range.end, Arrays.asList("matchedGroup", "group" + i));
            }
        }
        textArea.setStyle("-fx-fill: red;");
    }

    private class Highlighter extends EnglishCommandsBaseVisitor<Void> {
        @Override
        public Void visitDateTime(EnglishCommandsParser.DateTimeContext ctx) {
            markKeyWords(ctx);
            return null;
        }

        @Override
        public Void visitTimeFormat(EnglishCommandsParser.TimeFormatContext ctx) {
            super.visitTimeFormat(ctx);
            markKeyWords(ctx);
            return null;
        }

        private void markKeyWords(ParseTree ctx) {
            if (ctx instanceof TerminalNode) {
                Token token = ((TerminalNode) ctx).getSymbol();
                switch (token.getType()) {
                    case EnglishCommandsParser.SINCE:
                    case EnglishCommandsParser.FROM:
                    case EnglishCommandsParser.AGO:
                    case EnglishCommandsParser.AT:
                    case EnglishCommandsParser.TO:
                    case EnglishCommandsParser.UNTIL:
                    case EnglishCommandsParser.FIN:
                    case EnglishCommandsParser.RESUME:
                    case EnglishCommandsParser.LAST:
                        addHighlight(token, token, "keyword");
                        break;
                    case EnglishCommandsParser.DAYS:
                    case EnglishCommandsParser.HOURS:
                    case EnglishCommandsParser.MINUTES:
                    case EnglishCommandsParser.SECONDS:
                        addHighlight(token, token, "timeUnit");
                        break;
                    case EnglishCommandsParser.NUMBER:
                    case EnglishCommandsParser.COLON:
                        addHighlight(token, token, "value");
                        break;
                    default:
                        break;
                }
            }
            if (ctx instanceof EnglishCommandsParser.DateTimeContext) {
                EnglishCommandsParser.DateTimeContext dateTimeContext = (EnglishCommandsParser.DateTimeContext) ctx;
                addHighlight(dateTimeContext.start, dateTimeContext.stop, "dateTime");
            } else {
                for (int i = 0; i < ctx.getChildCount(); i++) {
                    markKeyWords(ctx.getChild(i));
                }
            }
        }

        private void addHighlight(Token start, Token stop, String style) {
            textArea.setStyle(start.getStartIndex(), stop.getStopIndex() + 1, Collections.singleton(style));
        }
    }

    public static class Factory {
        private final ItemGrouper itemGrouper;
        private final ReportConfig config;

        @Inject
        public Factory(ItemGrouper itemGrouper, ReportConfig config) {
            this.itemGrouper = itemGrouper;
            this.config = config;
        }

        public CommandHighlighter create(StyleClassedTextArea textArea) {
            return new CommandHighlighter(itemGrouper, config, textArea);
        }
    }
}

File: src/main/kotlin/org/stt/gui/jfx/text/CommandHighlighter.kt
package org.stt.gui.jfx.text

import org.antlr.v4.runtime.CommonTokenStream
import org.antlr.v4.runtime.Token
import org.antlr.v4.runtime.tree.ParseTree
import org.antlr.v4.runtime.tree.TerminalNode
import org.fxmisc.richtext.StyleClassedTextArea
import org.stt.command.CaseInsensitiveInputStream
import org.stt.config.ReportConfig
import org.stt.grammar.EnglishCommandsBaseVisitor
import org.stt.grammar.EnglishCommandsLexer
import org.stt.grammar.EnglishCommandsParser
import org.stt.text.ItemGrouper
import org.stt.text.Type
import java.util.*
import javax.inject.Inject

class CommandHighlighter(private val itemGrouper: ItemGrouper,
                         private val config: ReportConfig,
                         private val textArea: StyleClassedTextArea) {
    private val visitor = Highlighter()

    fun update() {
        if (textArea.length == 0) {
            return
        }
        textArea.clearStyle(0, textArea.length)
        val text = textArea.text
        val input = CaseInsensitiveInputStream(text)
        val lexer = EnglishCommandsLexer(input)
        val tokenStream = CommonTokenStream(lexer)
        val parser = EnglishCommandsParser(tokenStream)
        parser.command().accept(visitor)
        val groups = itemGrouper(text).toTypedArray()
        for (i in 0 until groups.size - 1) {
            val group = groups[i]
            if (group.type == Type.MATCH) {
                textArea.setStyle(group.range.start, group.range.end, Arrays.asList("matchedGroup", "group$i"))
            }
        }
        textArea.style = "-fx-fill: red;"
    }

    private inner class Highlighter : EnglishCommandsBaseVisitor<Void>() {
        override fun visitDateTime(ctx: EnglishCommandsParser.DateTimeContext): Void? {
            markKeyWords(ctx)
            return null
        }

        override fun visitTimeFormat(ctx: EnglishCommandsParser.TimeFormatContext): Void? {
            super.visitTimeFormat(ctx)
            markKeyWords(ctx)
            return null
        }

        private fun markKeyWords(ctx: ParseTree) {
            if (ctx is TerminalNode) {
                val token = ctx.symbol
                when (token.type) {
                    EnglishCommandsParser.SINCE, EnglishCommandsParser.FROM, EnglishCommandsParser.AGO, EnglishCommandsParser.AT, EnglishCommandsParser.TO, EnglishCommandsParser.UNTIL, EnglishCommandsParser.FIN, EnglishCommandsParser.RESUME, EnglishCommandsParser.LAST -> addHighlight(token, token, "keyword")
                    EnglishCommandsParser.DAYS, EnglishCommandsParser.HOURS, EnglishCommandsParser.MINUTES, EnglishCommandsParser.SECONDS -> addHighlight(token, token, "timeUnit")
                    EnglishCommandsParser.NUMBER, EnglishCommandsParser.COLON -> addHighlight(token, token, "value")
                    else -> Unit
                }
            }
            if (ctx is EnglishCommandsParser.DateTimeContext) {
                addHighlight(ctx.start, ctx.stop, "dateTime")
            } else {
                for (i in 0 until ctx.childCount) {
                    markKeyWords(ctx.getChild(i))
                }
            }
        }

        private fun addHighlight(start: Token, stop: Token, style: String) {
            textArea.setStyle(start.startIndex, stop.stopIndex + 1, setOf(style))
        }
    }

    class Factory @Inject
    constructor(private val itemGrouper: @JvmSuppressWildcards ItemGrouper, private val config: ReportConfig) {

        fun create(textArea: StyleClassedTextArea): CommandHighlighter {
            return CommandHighlighter(itemGrouper, config, textArea)
        }
    }
}
